### ISSUE TITLE
Warn on pointer cast with type inference

### ISSUE THREAD

**Issue by tiago-lqt** [rust-lang/rust-clippy#7028]
### What it does

Lint should warn when casts using type inference are being used on pointers types.

Recently we had a bug on a FFI boundary where a `*const T` was being cast with `self.inner() as *mut _` when calling a C function with signature `sys::foo(T**)`, but `inner` function is defined as `fn inner() -> *const T`.
This lead to passing a pointer as pointer to pointer, and the application crashing at runtime.

### Categories (optional)

- Kind: clippy::correctness, clippy::pedantic

Original code might be buggy. Intent is to have a mechanism that warns or errors when such a code is found on particular scenarios, specifically for our case in code involving the FFI boundary.

### Drawbacks

Might get annoying for most use cases, thats why it was suggested also as pedantic linter.

### Example

```rust
sys::foo(self.inner() as *mut _);
```

Could be written as:

```rust
sys::foo(self.inner() as *mut T);
```

In this case code would fail to compile, but provides us the chance to catch and fix the error at compile time.

**Comment by panosfol** [comment]
Any news on this? Im interested in working on this.

**Comment by Centri3** [comment]
There is already #10567 which implements this, but only the `*mut _` and `*const _` cases. Maybe a lint to detect any such cases could be nice.
